File: src/main/java/io/stouder/adonis/cli/json/ace/Command.java
package io.stouder.adonis.cli.json.ace;

import com.google.gson.*;
import io.stouder.adonis.cli.json.routes.ClosureRouteHandler;
import io.stouder.adonis.cli.json.routes.ControllerRouteHandler;
import io.stouder.adonis.cli.json.routes.RouteHandler;
import lombok.Getter;

import java.lang.reflect.Type;
import java.util.List;

@Getter
public class Command {
    private String commandName;
    private String description;
    private String namespace;
    private List<String> aliases;
    private List<CommandFlag> flags;
    private List<CommandArgument> args;
    private CommandOptions options;
    private String filePath;
    private String absolutePath;

    public static class Deserializer implements JsonDeserializer<Command> {
        @Override
        public Command deserialize(JsonElement jsonElement, Type type, JsonDeserializationContext jsonDeserializationContext) throws JsonParseException {
            JsonObject jsonObject = jsonElement.getAsJsonObject();
            JsonElement jsonType = jsonObject.get("help");
            if(jsonType.isJsonArray()) {
                return jsonDeserializationContext.deserialize(jsonElement, CommandComplexHelp.class);
            }
            return jsonDeserializationContext.deserialize(jsonElement, CommandSimpleHelp.class);
        }
    }
}

File: src/main/java/io/stouder/adonis/cli/json/ace/Command.kt
package io.stouder.adonis.cli.json.ace

import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import com.google.gson.annotations.SerializedName
import java.lang.reflect.Type

open class Command(
    @SerializedName("commandName") open val commandName: String,
    @SerializedName("description") open val description: String,
    @SerializedName("namespace") open val namespace: String,
    @SerializedName("aliases") open val aliases: List<String>,
    @SerializedName("flags") open val flags: List<CommandFlag<*>>,
    @SerializedName("args") open val args: List<CommandArgument>,
    @SerializedName("options") open val options: CommandOptions,
    @SerializedName("filePath") open val filePath: String,
    @SerializedName("absolutePath") open val absolutePath: String
) {
    class Deserializer : JsonDeserializer<Command> {
        @Throws(JsonParseException::class)
        override fun deserialize(jsonElement: JsonElement, type: Type, jsonDeserializationContext: JsonDeserializationContext): Command {
            val jsonObject = jsonElement.asJsonObject
            val jsonType = jsonObject.get("help")
            return if (jsonType.isJsonArray) {
                jsonDeserializationContext.deserialize(jsonElement, CommandComplexHelp::class.java)
            } else {
                jsonDeserializationContext.deserialize(jsonElement, CommandSimpleHelp::class.java)
            }
        }
    }
}